File: algorithms/rs/49.字母异位词分组.rs
/*
 * @lc app=leetcode.cn id=49 lang=rust
 *
 * [49] 字母异位词分组
 */

// @lc code=start
use std::collections::HashMap;
impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut map: HashMap<Vec<u8>, Vec<String>> = HashMap::new();
        for str in strs {
            let mut vec = str.as_bytes().to_vec();
            vec.sort();
            let mut vec = map.entry(vec).or_insert(vec![]);
            vec.push(str);
        }
        map.values().cloned().collect()
    }
}
// @lc code=end

File: algorithms/js/49.字母异位词分组.js
/*
 * @lc app=leetcode.cn id=49 lang=javascript
 *
 * [49] 字母异位词分组
 */

// @lc code=start
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
  const map = new Map();
  strs.forEach((str) => {
    const key = str
      .split("")
      .sort((a, b) => a.charCodeAt() - b.charCodeAt())
      .join("");
    map.set(key, map.has(key) ? [...map.get(key), str] : [str]);
  });
  return [...map.values()];
};
// @lc code=end
